### PR TITLE
safeguard againts LUA and module mismatch

### PR DESCRIPTION
**Pull request by StonedDawg**
add ERROR screen and version info when LUA version and TX module version is not compatible to safeguard people from using the wrong LUA script and possibly enter wrong parameters.

### CODE REVIEW

File: src/lua/ELRS.lua [hunk 1]
```diff
@@ -295,13 +301,13 @@ local function processResp()
...
 		
-			if(data[3] == 0xFF) and #data == 11 then
+			if(data[3] == 0xFF) and #data == 12 then
 				bindmode = bit32.btest(0x01, data[4]) -- bind mode active 
 				wifiupdatemode = bit32.btest(0x02, data[4]) 
-				
 				if StopUpdate == false then 
 					TLMinterval.selected = GetIndexOf(TLMinterval.values,data[6])
 					MaxPower.selected = GetIndexOf(MaxPower.values,data[7])
+                    tx_lua_version = data[12];
```
Comment (cruwaller) [inline src/lua/ELRS.lua:310]: Fix the code indentation.

File: src/lua/ELRS.lua [hunk 2]
```diff
@@ -219,27 +220,32 @@ local function refreshLCD()
...
-        local item_y = yOffset + offsets.top + radio_data.yOffset * item.index
-        if item.action ~= nil or item.func ~= nil then
-            lcd.drawText(lOffset, item_y, item.name, getFlags(idx) + radio_data.textSize)
-        else
-            local value = '?'
-			if 0 < item.selected and item.selected <= #item.list and gotFirstResp then
-            --if 0 < item.selected and item.selected <= #item.list and item.selected <= item.max_allowed then
-                value = item.list[item.selected]
+    if tx_lua_version == version then
+    yOffset = radio_data.yOffset_val
```
Comment (cruwaller) [inline src/lua/ELRS.lua:225]: indentation

**Review by cruwaller — CHANGES_REQUESTED** [review]
Thanks! 
This is good improvement.

**Comment by StonedDawg** [comment]
@cruwaller some parts of the code's indentation are spaces, and some parts are tabs, would you like me to change them to tabs?

**Comment by StonedDawg** [comment]
> Thanks!
> This is good improvement.

thanks! all fixed..

**Review by cruwaller — APPROVED** [review]

**Comment by CapnBry** [comment]
I just tested this on my develop branch and it works and is a good idea. I have two issues with it though:
* The updated lua running against an old TX that doesn't put out the version number reports "Update ELRS.lua to v.0(current v.3)". Maybe update refreshLCD() to point to the right direction. `if (tx_lua_version < version) then print("Update TX") else print("Update ELRS.lua") end` like that but prettier.
* If someone updates their TX and RX and heads out to the field without updating the LUA they're now 100% fully dead in the water when it comes to changing even the most basic ELRS parameters. Wouldn't it be better to gracefully degrade somehow? I could see this being a giant gotcha for something so active in development. Maybe have a minimum supported TX version and just lock out functionality based on tx_lua_version vs version. If they don't match draw a big inverse bar letting the user know which should be upgraded. I don't want the lua to be full of compatibility hacks but the way it is now means a dev just changing the version number locks everyone out and this is something that could be forgotten.

Pedantically, I'd also suggest renaming version to VERSION like it is a const/define so it is less ambiguous, or even LUA_PROTO_VERSION. Visually, maybe drop the decimal in the version number display too. v.0? v.3? why. Just v3

Note that I am not a developer so maybe take these with a grain of salt.

**Comment by StonedDawg** [comment]
> I just tested this on my develop branch and it works and is a good idea. I have two issues with it though:
> 
> * The updated lua running against an old TX that doesn't put out the version number reports "Update ELRS.lua to v.0(current v.3)". Maybe update refreshLCD() to point to the right direction. `if (tx_lua_version < version) then print("Update TX") else print("Update ELRS.lua") end` like that but prettier.
> * If someone updates their TX and RX and heads out to the field without updating the LUA they're now 100% fully dead in the water when it comes to changing even the most basic ELRS parameters. Wouldn't it be better to gracefully degrade somehow? I could see this being a giant gotcha for something so active in development. Maybe have a minimum supported TX version and just lock out functionality based on tx_lua_version vs version. If they don't match draw a big inverse bar letting the user know which should be upgraded. I don't want the lua to be full of compatibility hacks but the way it is now means a dev just changing the version number locks everyone out and this is something that could be forgotten.
> 
> Pedantically, I'd also suggest renaming version to VERSION like it is a const/define so it is less ambiguous, or even LUA_PROTO_VERSION. Visually, maybe drop the decimal in the version number display too. v.0? v.3? why. Just v3
> 
> Note that I am not a developer so maybe take these with a grain of salt.

hmm with the 1st issue, I think displaying its version is kinda important and might help in-the-field LUA matching between multiple firmwares (if elrs.lua has been saved in handset as elrsv1.lua, elrsv2.lua), it might help testing multiple PR on site. although words selection is also up to devs, I will change to fit the best outcome.

with the 2nd issue, I can think of another way, which is with "OK" to confirm and keep using LUA as it is even outdated. since the UART byte length between v2 and v3 is different, LUA would only shows "???", I can make a workaround to adjust it but then Im not sure if it would be considered as "neat" for the codes to do that... anyway its up to the devs? I will change to get the best outcome.

**Comment by CapnBry** [comment]
> hmm with the 1st issue, I think displaying its version is kinda important and might help in-the-field LUA matching between multiple firmwares (if elrs.lua has been saved in handset as elrsv1.lua, elrsv2.lua), it

Oh for sure yeah. I didn't type the whole thing out but I agree that showing the version number is important. My only complaint is that the way it is now always blames the LUA for being out of date even if it is the TX that is out of date.

**Comment by StonedDawg** [comment]
@CapnBry ahhh got it, how about we use the word "match" instead of update?

**Comment by CapnBry** [comment]
I was thinking something like this general idea
```
        lcd.drawText(lOffset, (radio_data.yOffset*2), "!!! VERSION MISMATCH !!!", INVERS)
        if (tx_lua_version > version) then
            lcd.drawText(lOffset, (radio_data.yOffset*3), "Update ELRS.lua", INVERS)
        else
            lcd.drawText(lOffset, (radio_data.yOffset*3), "Update TX module", INVERS)
        lcd.drawText(lOffset, (radio_data.yOffset*4), "LUA v"..version..", TX v"..tx_lua_version, INVERS)
```

**Comment by JyeSmith** [comment]
@StonedDawg this is good to merge when you are ready 👍 

I am happy either way with the wording and have a feeling it may change again once users start seeing the message.

**Review by JyeSmith — APPROVED** [review]

**Comment by StonedDawg** [comment]
@JyeSmith ok, I will changes words and add force use

**Comment by StonedDawg** [comment]
@JyeSmith ok its ready, it can be forced backward, but not to v01 and v02 because of different packet length, it will stuck on connecting when it happened

**Comment by StonedDawg** [comment]
> @JyeSmith ok its ready, it can be forced backward, but not to v01 and v02 because of different packet length, it will stuck on connecting when it happened

force use now will not check data length, this will be backward compatible with v02 but still not with v01 (v01 data value is out of range).

**Pull request merged**